wiki/development/coding-guidelines.md

```diff
@@ -8,27 +8,185 @@ order: 1
+
+Note that you need the strings in module stringtable.xml in the correct format
+`STR_ACE_<module>_<string>`</br>
+Example:</br>
+`STR_Balls_Banana`</br>
+
+Script strings:
+
+| Macro  |  Expands to |
+| -------|---------|
+| `LSTRING(banana)` |  "STR_ACE_balls_banana"|
+| `ELSTRING(balls,banana)` | "STR_ACE_balls_banana"|
+
+
+Config Strings (require `$` as first character):
+
+| Macro  |  Expands to |
+| -------|---------|
+| `CSTRING(banana)` |  "$STR_ACE_balls_banana" |
+| `ECSTRING(balls,banana)` | "$STR_ACE_balls_banana" |
+
+
+## 3. Functions
+
+Functions should be created in the functions\ subdirectory, named `fnc_functionName.sqf` They should then be indexed via the `PREP(functionName)` macro in the XEH_preInit.sqf file.
+
+The `PREP` macro allows for CBA function caching, which drastically speeds up load times. **Beware though that function caching is enabled by default and as such to disable it you need to `#define DISABLE_COMPILE_CACHE` above your `#include "script_components.hpp"` include!**
+
+### 3.1. Headers
+Every function should have a header of the following format as the start of their function file:
+
+```cpp
+/*
+ * Author: [Name of Author(s)]
+ * [Description]
+ *
+ * Arguments:
+ * 0: The first argument <STRING>
+ * 1: The second argument <OBJECT>
+ *
+ * Return Value:
+ * The return value <BOOL>
+ *
+ * Example:
+ * ["something", player] call ace_common_fnc_imanexample
+ *
+ * Public: [Yes/No]
+ */
+```
+
+This is not the case for inline functions or functions not containing their own file.
+
+### 3.2. Includes
+Every function includes the `script_component.hpp` file just below the function header. Any additional includes or defines must be below this include.
+
+All scripts written must be below this include and any potential additional incldues or defines.
 
-Tabs can be tricky sometimes, especially when it comes to sharing code with others. Additionally, a lot of people tend to forget they're using tabs when they're aligning things after the first character, which causes things to fall apart when viewing the code at different tab lengths.
+#### 3.2.1. Reasoning
+This ensures every function starts of in an uniform way and enforces function documentation.
 
+## 4. Global Variables
+All Global Variables are defined in the XEH_preInit.sqf file of the component they will be used in with an initial default value.
 
-## 2. Braces
+Exceptions:
+* Dynamically generated global variables
+* Variables that do not origin from he ACE project, such as BI global variables or third party such as CBA.
 
-- opening brace on the same line as keyword
-- closing brace in own line, same level of indentation as keyword
+## 5. Code Style
+
+### 5.1. Braces placement
+Braces "{ }" which enclose a code block will have the first bracket placed behind the statement in case of if, switch statements or while, waituntil & for loops. The second bracket will be placed on the same column as the statement but on a separate line.
```

Comment:
Braces `{ }` which enclose a code block will have the first bracket placed behind the statement in case of an `if`, `switch`, `while`, `waitUntil` statements and for loops. The second brace will be placed on the same column as the statement but on a separate line.